python/plugins/processing/core/parameters.py

```diff
@@ -79,13 +80,25 @@ def setValue(self, obj):
         of parameter.
         """
         if obj is None:
-            if not self.optional:
+            if not self.optional and self.default is None:
                 return False
-            self.value = None
-            return True
+            elif self.default is None:
+                self.value = None
+                return True
+            else:
```

Comment:
@volaya Is there a reason that for this assumption? I see that it works, but protecting `value` from ever being assigned prohibited values sounds like a good idea (I think that's what @michaelkirk was pointing to). I don't see any disadvantages from doing so, but on the other hand it makes the code stronger.

I guess it's mostly about OO paradigms and encapsulation of variables :)